src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.SymbolToProjectId.cs

```diff
@@ -0,0 +1,136 @@
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class SolutionState
+    {
+        /// <summary>
+        /// Contains caches to help us quickly and accurately map from <see cref="ISymbol"/>s to <see
+        /// cref="ProjectId"/>.
+        /// </summary>
+        /// <remarks>
+        /// Note: this is not <see langword="readonly"/> as this is a struct and will mutate as operations are called on
+        /// it.  However, in the common case where this data is not needed, it will have not have any impact.
+        /// </remarks>
+        private SymbolToProjectId _symbolToProjectId;
+
+        /// <inheritdoc cref="Solution.GetExactProjectId"/>
+        public ProjectId? GetExactProjectId(ISymbol symbol)
+            => _symbolToProjectId.GetExactProjectId(symbol);
+
+        /// <summary>
+        /// A helper type that can be used to map from symbols to a project they could have come from.
+        /// </summary>
+        private struct SymbolToProjectId
+        {
+            /// <summary>
+            /// Weak cache from a <see cref="Compilation"/> to the ID of the <see cref="Project"/> it was created from.
+            /// Only used for mapping a <see cref="Project"/>'s <see cref="Compilation.GlobalNamespace"/> back.  These
+            /// <see cref="INamespaceSymbol"/>'s do not belong to a <see cref="IAssemblySymbol"/> or <see
+            /// cref="IModuleSymbol"/> and thus cannot be tracked with <see
+            /// cref="CompilationTracker.State.CompilationAssembliesAndModules"/>.
+            /// </summary>
+            private static readonly ConditionalWeakTable<Compilation, ProjectId> s_compilationToProjectId
+                = new ConditionalWeakTable<Compilation, ProjectId>();
+
+            private readonly SolutionState _solutionState;
+
+            /// <summary>
+            /// Cache we use to map between assembly and module symbols and the project they came from.  That way if we
+            /// are asked about many symbols from the same assembly/module we can answer the question quickly after
+            /// computing for the first one.
+            /// </summary>
+            private ConditionalWeakTable<ISymbol, ProjectId?>? _assemblyOrModuleSymbolToProjectId;
+            private static readonly Func<ConditionalWeakTable<ISymbol, ProjectId?>> s_createTable = () => new ConditionalWeakTable<ISymbol, ProjectId?>();
+
+            public SymbolToProjectId(SolutionState solutionState)
+            {
+                _solutionState = solutionState;
+                _assemblyOrModuleSymbolToProjectId = null;
+            }
+
+            /// <inheritdoc cref="Solution.GetExactProjectId(ISymbol)"/>.
+            public ProjectId? GetExactProjectId(ISymbol? symbol)
+            {
+                LazyInitialization.EnsureInitialized(ref _assemblyOrModuleSymbolToProjectId, s_createTable);
```

Comment:
initialized on demand so we don't pay any heap price for this information unless we have an operaiton that actually needs it.